include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -0,0 +1,683 @@
+#  include <deal.II/differentiation/ad/sacado_product_types.h>
+
+#  include <algorithm>
+#  include <iostream>
+#  include <iterator>
+#  include <numeric>
+#  include <set>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Differentiation
+{
+  namespace AD
+  {
+    /**
+     * A base helper class that facilitates the evaluation of the derivative(s)
+     * of a number of user-defined dependent variables $\mathbf{f}(\mathbf{X})$
+     * with respect to a set of independent variables $\mathbf{X}$, that is
+     * $\dfrac{d^{i} \mathbf{f}(\mathbf{X})}{d^{i} \mathbf{X}}$.
+     *
+     * In addition to the dimension @p dim, this class is templated on the
+     * floating point type @p scalar_type of the number that we'd like to
+     * differentiate, as well as an enumeration indicating the @p ADNumberTypeCode .
+     * The @p ADNumberTypeCode dictates which auto-differentiation library is
+     * to be used, and what the nature of the underlying auto-differentiable
+     * number is. Refer to the @ref auto_symb_diff module for more details in
+     * this regard.
+     *
+     * For all of the classes derived from this base class, there are two
+     * possible ways that the code in which they are used can be structured.
+     * The one approach is effectively a subset of the other, and which might
+     * be necessary to use depends on the nature of the chosen
+     * auto-differentiable number.
+     *
+     * When "tapeless" numbers are employed, the most simple code structure
+     * would be of the following form:
+     *
+     * @code
+     *   // Initialize AD helper
+     *   ADHelperType<dim,tapeless_AD_typecode> ad_helper (...);
+     *
+     *   // Register independent variables
+     *   ad_helper.register_independent_variable(...);
+     *
+     *   // Extract the sensitive equivalent of the independent variables
+     *   // The operations conducted with these numbers will be tracked.
+     *   const auto ad_independent_variables
+     *     = ad_helper.get_sensitive_variables(...);
+     *
+     *   // Use the sensitive variables to compute the dependent variables
+     *   const auto ad_dependent_variables = func(sensitive_variables);
+     *
+     *   // Register the dependent variables with the helper class
+     *   ad_helper.register_dependent_variables(ad_dependent_variables);
+     *
+     *   // Compute derivatives of the dependent variables
+     *   const auto derivatives = ad_helper.compute_gradients();
+     * @endcode
+     *
+     * Note that since the specialized classes interpret the independent
+     * variables in different ways, above represents only an outline of the
+     * steps taken to compute derivatives. More specific examples are outlined
+     * in the individual classes that specialize this base class.
+     *
+     * When "taped" numbers are to be used, the above code should be wrapped by
+     * a few more lines of code to manage the taping procedure:
```

Comment:
I'm not enough of an expert to define what "taped" actually means. Maybe that too could become a glossary entry? My intuition is that you can use a tape to record the operations done only once, and so don't have to compute all of the derivatives every time you enter a function, but can reuse what's been done in a previous invokation. Is this about right?

Reply:
Yes, lets add glossary entries for these. But maybe I could do it as a follow up PR (see https://github.com/dealii/dealii/issues/7120).

Reply:
BTW. I think I give a pretty detailed description of "taped" in the [AD module](https://dealii.org/developer/doxygen/deal.II/group__auto__symb__diff.html). But this should still be added to a glossary entry.

Reply:
Separate PR is totally fine.

If you want to, you can also just link to that module -- maybe that's the better place anyway, since it is the one central place where everything AD related comes together. Either way is fine by me, though.